include/deal.II/fe/mapping_manifold.h

```diff
@@ -0,0 +1,581 @@
+     * function calls initialize().
+     */
+    void
+    initialize_face (const UpdateFlags      update_flags,
+                     const Quadrature<dim> &quadrature,
+                     const unsigned int     n_original_q_points);
+
+
+    /**
+     * Compute the weights associated to the Manifold object, that
+     * need to be passed when computing the location of the quadrature
+     * points.
+     */
+    void
+    compute_manifold_quadrature_weights(const Quadrature<dim> &quadrature);
+
+    /**
+     * Store vertices internally.
+     */
+    void
+    store_vertices(const typename Triangulation<dim,spacedim>::cell_iterator &cell) const;
+
+    /**
+     * Return an estimate (in bytes) or the memory consumption of this object.
+     */
+    virtual std::size_t memory_consumption () const;
+
+    /**
+     * The current cell vertices.
+     *
+     * Computed each.
+     */
+    mutable std::vector<Point<spacedim> > vertices;
+
+    /**
+     * The current cell.
+     *
+     * Computed each.
+     */
+    mutable typename Triangulation<dim,spacedim>::cell_iterator cell;
+
+    /**
+     * The actual quadrature on the reference cell.
+     *
+     * Computed once.
+     */
+    Quadrature<dim> quad;
+
+
+    /**
+     * Values of quadrature weights for manifold quadrature formulas.
+     *
+     * Computed once.
+     */
+    std::vector<std::vector<double> > cell_manifold_quadrature_weights;
+
+    // /**
+    //  * Values of shape function derivatives. Access by function @p derivative.
+    //  *
+    //  * Computed once.
+    //  */
+    // std::vector<Tensor<1,dim> > shape_derivatives;
+
+    // /**
+    //  * Values of shape function second derivatives. Access by function @p
+    //  * second_derivative.
+    //  *
+    //  * Computed once.
+    //  */
+    // std::vector<Tensor<2,dim> > shape_second_derivatives;
+
+    // /**
+    //  * Values of shape function third derivatives. Access by function @p
+    //  * second_derivative.
+    //  *
+    //  * Computed once.
+    //  */
+    // std::vector<Tensor<3,dim> > shape_third_derivatives;
+
+    // /**
+    //  * Values of shape function fourth derivatives. Access by function @p
+    //  * second_derivative.
+    //  *
+    //  * Computed once.
+    //  */
+    // std::vector<Tensor<4,dim> > shape_fourth_derivatives;
+
+    /**
+     * Unit tangential vectors. Used for the computation of boundary forms and
+     * normal vectors.
+     *
+     * This vector has (dim-1)GeometryInfo::faces_per_cell entries. The first
+     * GeometryInfo::faces_per_cell contain the vectors in the first
+     * tangential direction for each face; the second set of
+     * GeometryInfo::faces_per_cell entries contain the vectors in the second
+     * tangential direction (only in 3d, since there we have 2 tangential
+     * directions per face), etc.
+     *
+     * Filled once.
+     */
+    std::vector<std::vector<Tensor<1,dim> > > unit_tangentials;
```

Comment:
Since you know the size of the array, make it a fixed size array, via either
```
  std::vector<Tensor<1,dim> > unit_tangentials[dim-1];
```
or
```
  std_cxx11::array<dim-1,std::vector<Tensor<1,dim> > > unit_tangentials;
```
This also saves you the code for sizing the array correctly.